lib/galaxy/webapps/galaxy/services/datasets.py

```diff
@@ -483,6 +515,29 @@ def converted(
         hda = self.hda_manager.get_accessible(decoded_id, trans.user)
         return self.hda_serializer.serialize_converted_datasets(hda, 'converted')
 
+    def delete_batch(
+        self,
+        trans: ProvidesHistoryContext,
+        payload: DeleteDatasetBatchPayload,
+    ) -> DeleteDatasetBatchResult:
+        success_count = 0
+        errors = []
+        for encoded_id in payload.ids:
```

Comment:
I am worried about users "destroying" their dataset collections by deleting members of their dataset collections here. That is a hard problem unfortunately. Maybe we need to split the HDAs that can be deleted into those that appear as DatasetCollectionElements and those that do not ?

Reply:
Ah, and we should also treat HDAs that appear in published histories differently (probably in the listing step, not here. If you send the item id here we should just assume you know what you're doing).

Reply:
I agree these backend changes are a bit simplistic just to get something to test with. I will open separate PRs for them and include your comments. Then we can iterate over those PRs and drop the API changes from this one once they are merged.

Reply:
I don't particularly mind having the backend changes in here, if it's not a huge amount of refactoring it helps to see why it was changed.

Reply:
Brainstorming.. Unless this has been implemented already (looking at the model, I don't see it), we could add a field to the HDA model that stores a count of its parent collections. Whenever an HDA instance is added to a collection, we increment that field. Deleting a collection decrements the counters of its members. Deleting an HDA is only allowed when the field is 0. Like ref counts in garbage collection.

Reply:
That would be a field that potentially sees a lot of concurrent updates for a usecase that is probably not run that often (i suppose you're not constantly hitting the limit / worried about your storage after all).  it's not that hard to generate a query for this. If  a query turns out to be a problem we can still think about alternatives.

Reply:
Oh yes, you're right. And now that I think of it, a query shouldn't be expensive.

Reply:
So, to recap and see if I understood correctly, this `delete_batch` datasets endpoint **should assume the caller knows what she/he is doing**. And is **the job of the `fetchItems` query to filter out unwanted datasets**, like "public" or those "part of a collection".
We can always add specific cleanup operations extensions targeting collections or entire histories.

Reply:
There are pros and cons to doing it this way.
The downside is that in between fetchItems and calling `delete_batch` the user may have changed something (published a history, added a HDA to a collection) and now their deleting things that should have probably not been deleted ... but that doesn't seem like a very likely scenario.

The pro is that it is simpler and a good enough start IMO.
One way to give stronger guarantees about not deleting unwanted items is to build an atomic update query where you basically use the same query that you'd use to get a list of items that is not in a published history / part of a collection and you turn that into the update statement. That would probably be a significant amount of additional effort.

My gut feeling is that as long as we document in the delete_batch endpoint that no additional checks (beyond permission) are performed that this is good enough, and we could add a "strict" mode later if it becomes necessary.

Reply:
(but if you want to implement the "strict" mode or another way to do something like this now that's cool too, it is your call)